service/controller/release/resource/status/cluster.go

```diff
@@ -0,0 +1,205 @@
+package status
+
+import (
+	"context"
+
+	apiexlabels "github.com/giantswarm/apiextensions/v2/pkg/label"
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	azurecapi "sigs.k8s.io/cluster-api-provider-azure/api/v1alpha3"
+
+	"github.com/giantswarm/release-operator/service/controller/key"
+)
+
+type TenantCluster struct {
+	ID               string
+	OperatorVersion  string
+	ProviderOperator string
+	ReleaseVersion   string
+}
+
+// Takes a list of tenant clusters and returns two maps containing the versions of their release and operator versions.
+func consolidateClusterVersions(clusters []TenantCluster) (map[string]bool, map[string]map[string]bool) {
+	releaseVersions := make(map[string]bool)
+	operatorVersions := make(map[string]map[string]bool)
+
+	// operatorVersions is a nested map including the operator name and version
+	// e.g. operatorVersions["aws-operator"]["8.7.6"]:true
+
+	for _, c := range clusters {
+		releaseVersions[c.ReleaseVersion] = true
+
+		if operatorVersions[c.ProviderOperator] == nil {
+			operatorVersions[c.ProviderOperator] = make(map[string]bool)
+		}
+		operatorVersions[c.ProviderOperator][c.OperatorVersion] = true
+	}
+
+	return releaseVersions, operatorVersions
+}
+
+// Returns a list of tenant clusters currently running on the installation.
+func (r *Resource) getCurrentTenantClusters(ctx context.Context) ([]TenantCluster, error) {
+
+	var tenantClusters []TenantCluster
+	{
+		awsClusters, err := r.getCurrentAWSClusters(ctx)
+		if IsResourceNotFound(err) {
+			// Fall through
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		tenantClusters = append(tenantClusters, awsClusters...)
+
+		azureClusters, err := r.getCurrentAzureClusters(ctx)
+		if IsResourceNotFound(err) || IsNoMatchesForKind(err) {
+			// Fall through
+		} else if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		tenantClusters = append(tenantClusters, azureClusters...)
+
+		legacyClusters, err := r.getLegacyClusters(ctx)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		tenantClusters = append(tenantClusters, legacyClusters...)
+	}
+
+	return tenantClusters, nil
+}
+
+// Returns a list of AWS clusters according to the awscluster resource.
+func (r *Resource) getCurrentAWSClusters(ctx context.Context) ([]TenantCluster, error) {
+	awsclusters, err := r.k8sClient.G8sClient().InfrastructureV1alpha2().AWSClusters("default").List(ctx, metav1.ListOptions{})
```

Comment:
I think Firecracker is working on adding multi-namespace support for AWSClusters. Might be worth checking with them how far out this is.

Reply:
Turns out the solution for this also addresses your comment about the `CtrlClient` -- this will search across namespaces now and is the same structure as for Azure